client/service/src/status_sinks.rs

```diff
@@ -14,19 +14,27 @@
 // You should have received a copy of the GNU General Public License
```

Comment:
How about following the `futures::sink::Sink` pattern for `StatusSinks`.

`StatusSinks` would have two states, initialized with the latter:

1. `Ready(YieldAfter)`

2. `Pending`

A user would

1. Call `poll_ready`. In case a outbound sender's interval has fired it would move the `StatusSinks` to status `Ready(YieldAfter)` and return `Poll::Ready()`.

2. Call `start_send`. It would take the `YieldAfter` from the `Ready` status, send the status down the channel and move the `StatusSink` status back to `Pending`.

I would find this easier to parse, not because the current implementation is complex or complicated, but because `futures::sink::Sink` is a known familiar pattern. As a nice-to-have it would relief us from the `entries_{tx,rx}` channel. If I am not mistaken none of this is performance critical, thus I don't understand the pattern of only locking `inner` within `next`.

Reply:
> I don't understand the pattern of only locking inner within next.

The reason is that we are sharing an `Arc<StatusSinks>` throughout in the networking code.

Calling `next()` returns a `Future` that holds a lock to the queue until one element is ready. If we locked in `push` as well, then as long as this `Future` is alive (which is 99%), then calling `status_sink.push()` would need to block as well.

There is no way you can get rid of the `entries_tx/rx` channel without making the `NetworkService::event_stream()` function blocking or asynchronous
(except by replacing it by something vaguely equivalent to a channel, like a Vec plus a task notifier).

Reply:
As for the `Sink` suggestion, we can't really do that.

We call `status_sink::next()` in the `NetworkWorker::next_action` async method. But if something else is ready (like the import queue, or a network message) then we drop the `Future` returned by `next()` and re-acquire it the next time `next_action` is called.

If we were to turn that into a `Sink`, then destroying that future would drop the element that we are trying to send.
It also means that the element that we send might already be obsolete when we actually send it.